JAVA-2649: Implement driver benchmark specification

jyemin opened the pull request:
See https://github.com/mongodb/specifications/tree/master/source/benchmarking
for the specification that this implements.

I suggest a quick read of the specification first for context, particularly around how long each benchmark executes (encapsulated in the BenchmarkRunner class) and the peculiarities of each individual benchmark.

The import/export benchmarks are the most complicated, as the specification encourages implementors to get creative about making them run fast.  I chose to do this by judicious use of concurrency.

jyemin commented on .evergreen/.evg.yml:
```diff
@@ -584,6 +605,13 @@ buildvariants:
   tasks:
      - name: "gssapi-auth-test"
 
+- matrix_name: "perf"
```
Currently this will run on every commit but if it's too much we can set batchtime

rozza replied:
Lets do that

jyemin replied:
Done

jyemin commented on .evergreen/run-perf-tests.sh:
```diff
@@ -0,0 +1,23 @@
+set -o errexit
+
+rm -rf driver-performance-test-data
+git clone https://github.com/mongodb-labs/driver-performance-test-data.git
```
Clone and expand the test data

jyemin commented on .evergreen/.evg.yml:
```diff
@@ -325,6 +325,21 @@ functions:
+          ${PREPARE_SHELL}
+          PROJECT_DIRECTORY=${PROJECT_DIRECTORY} .evergreen/run-perf-tests.sh
+
+  "send dashboard data":
```
This is the way to interact with the performance graphing feature of Evergreen.  The benchmark suite will write the src/results.json file to this location.

jyemin commented on driver-benchmarks/build.gradle:
```diff
@@ -0,0 +1,39 @@
+    compile 'ch.qos.logback:logback-classic:1.1.1'
+}
+
+applicationDefaultJvmArgs = ["-Dorg.mongodb.benchmarks.data=" + System.getProperty('org.mongodb.benchmarks.data'),
```
Chain the system properties from Gradle to the BenchmarkSuite that it's running.

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/BenchmarkSuite.java:
```diff
@@ -0,0 +1,127 @@
+    private static final List<BenchmarkResultWriter> WRITERS = Arrays.<BenchmarkResultWriter>asList(
+            new EvergreenBenchmarkResultWriter());
+
+    public static void main(String[] args) throws Exception {
```
This is the entry point, which runs all the benchmarks defined in the suite.

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/BenchmarkSuite.java:
```diff
@@ -0,0 +1,127 @@
+        if (args.length > 0) {
+            clazz = Class.forName(args[0]);
+        } else {
+            clazz = Document.class;
```
Currently we're just testing with Document.

jyemin replied:
Decided to remove all this and just use Document

jyemin commented on driver-benchmarks/build.gradle:
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2016 MongoDB, Inc.
```
Just noticed this (first version of this was done a long time ago...).  I'll fix up the copyrights.

jyemin replied:
Done

jyemin commented on settings.gradle:
```diff
@@ -14,5 +14,4 @@
 
-include 'util','bson', 'driver-core', 'driver-sync', 'driver-async', 'driver-legacy', 'mongodb-driver', 'driver-embedded', 'mongo-java-driver'
-
+include 'util','bson', 'driver-core', 'driver-sync', 'driver-async', 'driver-legacy', 'mongodb-driver', 'driver-embedded', 'mongo-java-driver', 'driver-benchmarks'
```
I'll add a newline

jyemin replied:
Done

rozza commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/GridFSImportBenchmark.java:
```diff
@@ -0,0 +1,107 @@
+    public void setUp() throws Exception {
+        super.setUp();
+        database = client.getDatabase("perftest");
+        bucket = GridFSBuckets.create(database);
```
Do we have to use the default chunksize? - I imagine it will have a large impact on the results for GridFS.

jyemin replied:
I don't think we do.  What would you recommend?

jyemin replied:
Done

rozza commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/GridFSExportBenchmark.java:
```diff
@@ -0,0 +1,167 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class GridFSExportBenchmark extends AbstractMongoBenchmark {
```
Suggest all benchmarks names are aligned to the `getName()` value.

jyemin replied:
Done

rozza commented on driver-benchmarks/src/main/com/mongodb/benchmark/framework/Benchmark.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private String getResourceRoot() {
+        // TODO: remove default
```
This needed anymore?

jyemin replied:
Done

rozza commented on driver-benchmarks/src/main/com/mongodb/benchmark/framework/EvergreenBenchmarkResultWriter.java:
```diff
@@ -0,0 +1,74 @@
+        jsonWriter.writeEndDocument();
+        jsonWriter.close();
+        writer.close();
+        System.out.println("DONE");
```
Remove?

jyemin replied:
It's useful when running in Evergreen as it's the only way to know which benchmark is hanging.

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/AbstractMongoBenchmark.java:
```diff
@@ -0,0 +1,37 @@
+    protected MongoClient client;
+
+    public void setUp() throws Exception {
+       client = MongoClients.create();
```
We only plan to test against a local standalone.  OK to leave this as is until we change our mind?

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/BenchmarkSuite.java:
```diff
@@ -0,0 +1,127 @@
+            };
+        }
+
+        throw new UnsupportedOperationException("Pick a different class");
```
I'm just going to remove support for anything but Document in this class.  It's just confusing.

jyemin replied:
Done

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/ExportBenchmark.java:
```diff
@@ -0,0 +1,207 @@
+                    try {
+                        RawBsonDocumentCodec codec = new RawBsonDocumentCodec();
+                        for (RawBsonDocument cur : documents) {
+                            codec.encode(new JsonWriter(writer), cur, EncoderContext.builder().build());
```
The default is to not indent.

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/GridFSExportBenchmark.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    public static void main(String[] args) throws Exception {
+        BenchmarkResult benchmarkResult = new BenchmarkRunner(new GridFSExportBenchmark(), 20, 100).run();
```
The main methods are not actually used except for debugging a single benchmark.  I can remove them all if that's confusing.

rozza replied:
Ah I that makes more sense!

jyemin commented:
@rozza commented back. Anything without a comment I've fixed and will push along with the resolutions to the currently open questions.

jyemin commented on build.gradle:
```diff
@@ -131,7 +131,7 @@ def getGitVersion() {
 }
 
-configure(subprojects.findAll { it.name != 'util' && it.name != 'mongo-java-driver' }) {
+configure(subprojects.findAll { it.name != 'util' && it.name != 'mongo-java-driver' && it.name != 'driver-benchmarks' }) {
```
Done

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/AbstractFindBenchmark.java:
```diff
@@ -0,0 +1,94 @@
+
+    public void setUp() throws Exception {
+        super.setUp();
+        MongoDatabase database = client.getDatabase("perftest");
```
Done

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/BenchmarkSuite.java:
```diff
@@ -0,0 +1,127 @@
+
+    private static void runBenchmark(final Benchmark benchmark) throws Exception {
+        long startTime = System.currentTimeMillis();
+        BenchmarkResult benchmarkResult = new BenchmarkRunner(benchmark, NUM_WARMUP_ITERATIONS, NUM_ITERATIONS, 60, 300 ).run();
```
Done

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/BenchmarkSuite.java:
```diff
@@ -0,0 +1,127 @@
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class BenchmarkSuite {
+
+    public static final int NUM_WARMUP_ITERATIONS = 1;
```
Done

jyemin commented on driver-benchmarks/src/main/com/mongodb/benchmark/benchmarks/GridFSExportBenchmark.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    public String getName() {
+        return "GridFS multi-file upload";
```
Done

jyemin commented:
Patch build: https://evergreen.mongodb.com/version/5b58822ae3c33108d40d6d8c

jyemin commented:
* Composite scores seem silly now and I think a bit painful to implement given how the benchmarks are abstracted.
* Added statics for warmups and iterations
* Added statics for thread pool sizes.  It's been a long time, but as I recall I arrived at those size via trial and error to get the best performance.  Since we're not measuring drivers against each other though, this doesn't matter so much so if you'd like I can centralize them.

jyemin commented:
Thread pool sizes are centralized.
